File: python/predict.py
#!/usr/bin/python3

""" Sample predict call showing how to pass event

Customize with placement, event type as needed

"""

import datetime

from google.cloud import retail
from google.oauth2 import service_account
from google.protobuf.timestamp_pb2 import Timestamp

SERVICE_ACCOUNT_FILE = "<path to service account json file>"
PROJECT_NUM = "<project # or id>"

SCOPES = ['https://www.googleapis.com/auth/cloud-platform']
credentials = service_account.Credentials.from_service_account_file(
  SERVICE_ACCOUNT_FILE, scopes=SCOPES)

client = retail.PredictionServiceClient(credentials=credentials)

timestamp = Timestamp()
timestamp.FromDatetime(dt=datetime.datetime.now())

user_event = {
  "event_type": "detail-page-view",
  "event_time": timestamp,
  "visitor_id": "visitor-1",
  "user_info": {
    "user_id": "user12345"
  },
  "product_details": [{
    "product": {
      "id": "12345"
    }
  }]
}

request = {
  "placement":
    'projects/' + PROJECT_NUM +
    '/locations/global/catalogs/default_catalog/placements/home_page',
  
  "user_event": user_event,

  "filter": "filterOutOfStockItems",

  "params": {
    "returnProduct": True,
    "returnScore": True
  }
}

response = client.predict(request)

print(response)

File: php/predict.php
<?php

// Example using Retail php client libraries to get prediction results

require 'vendor/autoload.php';

use Google\Cloud\Retail\V2\PredictionServiceClient;
use Google\Cloud\Retail\V2\PredictResponse;
use Google\Cloud\Retail\V2\Product;
use Google\Cloud\Retail\V2\ProductDetail;
use Google\Cloud\Retail\V2\UserEvent;

$predictionClient = new PredictionServiceClient([
  'credentials' => '[your service account key.json file]'
]);

$placement = "projects/[YOUR PROJECT #]/locations/global/catalogs/default_catalog/placements/product_detail";

$product_detail = new ProductDetail([
  'product' => new Product(['id' => '12345']),
  'quantity' => 1
]);

$userEvent = new UserEvent([
  'event_type' => 'detail-page-view',
  'visitor_id' => 'ABCDEFG',
  'product_details' => [
    new ProductDetail([
      'product' => new Product(['id' => '12345']),
      'quantity' => 1
    ])
  ]
]);

$pb_true = new Google\Protobuf\Value();
$pb_true->setBoolValue(true);

try {
  $predictions = $predictionClient->predict($placement, $userEvent, [
    'params' => ['returnProduct' => $pb_true],
    'filter' => 'filterOutOfStockItems'
    ]
  );
} finally {
  $predictionClient->close();
}

$results = $predictions->getResults();

print($results->count() . " Item Id's Returned:");

$iterator = $results->getIterator();
while($iterator->valid()) {
  print($iterator->current()->getId() . "\n");
  
  // With returnProduct=true we can get all the product data
  print_r(json_decode($iterator->current()->serializeToJsonString()));
  
  $iterator->next();
}

?>
